rewrite-java/src/main/java/org/openrewrite/java/AddOrUpdateAnnotationAttribute.java

```diff
@@ -236,33 +267,132 @@ public J.Annotation visitAnnotation(J.Annotation a, ExecutionContext ctx) {
+                                            .getArguments()).get(0);
+                                    jLiteralList.add(j, e);
+                                }
+                            }
+
+                            return arrayValue.withInitializer(jLiteralList);
                         }
                         return it;
                     });
+
                     if (newArgs != currentArgs) {
                         a = a.withArguments(newArgs);
                     }
-                    if (foundOrSetAttributeWithDesiredValue.get()) {
-                        a = maybeAutoFormat(original, a, ctx);
-                        return a;
+                    if (!foundOrSetAttributeWithDesiredValue.get() && !attributeValIsAlreadyPresent(newArgs, newAttributeValue)) {
+                        // There was no existing value to update, so add a new value into the argument list
+                        String effectiveName = (attributeName == null) ? "value" : attributeName;
+                        //noinspection ConstantConditions
+                        J.Assignment as = (J.Assignment) ((J.Annotation) JavaTemplate.builder("#{} = #{}")
+                                .contextSensitive()
+                                .build()
+                                .apply(getCursor(), a.getCoordinates().replaceArguments(), effectiveName, newAttributeValue))
+                                .getArguments().get(0);
+                        a = a.withArguments(ListUtils.concat(as, a.getArguments()));
                     }
-                    // There was no existing value to update, so add a new value into the argument list
-                    String effectiveName = (attributeName == null) ? "value" : attributeName;
-                    //noinspection ConstantConditions
-                    J.Assignment as = (J.Assignment) ((J.Annotation) JavaTemplate.builder("#{} = #{}")
-                            .contextSensitive()
-                            .build()
-                            .apply(getCursor(), a.getCoordinates().replaceArguments(), effectiveName, newAttributeValue))
-                            .getArguments().get(0);
-                    a = a.withArguments(ListUtils.concat(as, a.getArguments()));
-                    a = maybeAutoFormat(original, a, ctx);
                 }
-
+                a = maybeAutoFormat(original, a, ctx);
                 return a;
             }
         });
     }
 
+    private static boolean valueMatches(@Nullable Expression expression, @Nullable String oldAttributeValue) {
+        if (expression == null) {
+            return oldAttributeValue == null;
+        }
+        if (oldAttributeValue == null) { // null means wildcard
+            return true;
+        } else if (expression instanceof J.Literal) { // 219 & 203
+            return oldAttributeValue.equals(((J.Literal) expression).getValue());
+        } else if (expression instanceof J.FieldAccess) { // 236
```

Comment:
Perhaps I'm missing something, but what do these numbers in the comments refer to?

Reply:
Line numbers where this branch applies :D helped me to backtrack while debugging. will remove and merge